docs/development/go/websocket/index.md

```diff
@@ -0,0 +1,651 @@
+
+{{< note >}}
+This guide was written with Go version 1.14.
+{{< /note >}}
+
+As we are going to use the `gorilla/websocket` Go package, it would be good to download it on your local machine by executing the next command:
+
+    go get -u github.com/gorilla/websocket
+
+## The WebSocket Protocol
+
+This section will briefly discuss the WebSocket protocol and tell you why it is important.
+
+The WebSocket protocol is a computer communications protocol that provides *full-duplex* communication channels over a single TCP connection. The WebSocket Protocol is defined in [RFC 6455](https://tools.ietf.org/html/rfc6455) and uses `ws://` and `wss://` instead of `http://` and `https://`, respectively. Therefore, the client should begin a WebSocket connection by using a URL that begins with `ws://`.
+
+The advantages of the WebSocket Protocol include the following:
+
+- A WebSocket connection is a full-duplex, bidirectional communications channel.
+- WebSocket connections are raw TCP-Sockets, which means that they do not have the overhead required for establishing an HTTP connection.
+- WebSocket connections can also be used for sending HTTP data.
+- WebSocket connections live until they are killed.
+- WebSocket connections can be used for real-time web applications.
+- Data can be sent from server to client at any time, without the client even requesting it.
+- WebSockets are a part of the HTML5 specification, which means that they are supported by all modern Web browsers.
+
+## Developing a WebSocket Server in Go
+
+In this section we are going to develop a small yet fully functional WebSocket server in Go using [gorilla/websocket](https://github.com/gorilla/websocket) in order to be able to test the WebSocket clients that will be implemented later on. The server implements the *Echo service*, which means that it automatically returns its input back to the client.
+
+Before showing the server implementation, it would be good for you to know that the `websocket.Upgrader` method of the `gorilla/websocket` package upgrades an HTTP server connection to the WebSocket protocol and allows you to define the parameters of the upgrade. After that, your HTTP connection is a WebSocket connection, which means that you will not be allowed to execute statements that work with the HTTP protocol.
+
+### The WebSocket Server Implementation
+
+The contents of `WSserver.go` are the following:
+
+{{< file "./WSserver.go" go >}}
+package main
+
+import (
+  "fmt"
+  "log"
+  "net/http"
+  "os"
+  "time"
+
+  "github.com/gorilla/websocket"
+)
+
+var PORT = ":1234"
+
+var upgrader = websocket.Upgrader{
+  ReadBufferSize:  1024,
+  WriteBufferSize: 1024,
+}
+
+func rootHandler(w http.ResponseWriter, r *http.Request) {
+  fmt.Fprintf(w, "Welcome!\n")
+  fmt.Fprintf(w, "Please use /ws for WebSocket!")
+}
+
+func wsHandler(w http.ResponseWriter, r *http.Request) {
+  log.Println("Connection from:", r.Host)
+  upgrader.CheckOrigin = func(r *http.Request) bool {
```

Comment:
This line of code concurrently modifies package-level variable upgrader. Fix by setting the field once in the composite literal above.  You should also note the security implications of this function.

Reply:
Thanks!